src/backend/distributed/metadata/node_metadata.c

```diff
@@ -426,22 +428,64 @@ citus_disable_node(PG_FUNCTION_ARGS)
 {
 	text *nodeNameText = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
+	bool forceDisableNode = PG_GETARG_BOOL(2);
+
 	char *nodeName = text_to_cstring(nodeNameText);
 	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
-	bool isActive = false;
-	bool onlyConsiderActivePlacements = false;
 	MemoryContext savedContext = CurrentMemoryContext;
 
 	PG_TRY();
 	{
-		if (NodeIsPrimary(workerNode))
+		WorkerNode *firstWorkerNode = GetFirstPrimaryWorkerNode();
+		if (!forceDisableNode && firstWorkerNode &&
+			firstWorkerNode->nodeId == workerNode->nodeId)
 		{
 			/*
-			 * Delete reference table placements so they are not taken into account
-			 * for the check if there are placements after this.
+			 * We sync metadata async and optionally in the background worker,
+			 * it would mean that some nodes might get the updates while other
+			 * not. And, if the node metadata that is changing is the first
+			 * worker node, the problem gets nasty. We serialize modifications
+			 * to replicated tables by acquiring locks on the first worker node.
+			 *
+			 * If some nodes get the metadata changes and some do not, they'd be
+			 * acquiring the locks on different nodes. Hence, having the
+			 * possibility of diverged shard placements for the same shard.
+			 *
+			 * To prevent that, we currently do not allow disabling the first
+			 * worker node.
 			 */
-			DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("disabling the first worker node in the "
+								   "metadata is not allowed")));
+		}
+
+		/*
+		 * First, locally mark the node as inactive. We'll later trigger background
+		 * worker to sync the metadata changes to the relevant nodes.
+		 */
+		bool isActive = false;
+		workerNode =
+			SetWorkerColumnLocalOnly(workerNode,
+									 Anum_pg_dist_node_isactive,
+									 BoolGetDatum(isActive));
+		if (NodeIsPrimary(workerNode))
+		{
+			ErrorIfCoordinatorMetadataSetFalse(workerNode, BoolGetDatum(isActive),
+											   "isactive");
+
+			if (!forceDisableNode)
+			{
+				/*
+				 * We do not allow disabling nodes if it contains any
+				 * primary placement. This decision can be overriden by
+				 * the user such as when the node is down and there is no
+				 * other way.
+				 */
+				ErrorIfNodeContainsNonRemovablePlacements(workerNode);
```

Comment:
I don't really like the dual meaning of "force" as both "delete last shard placement" and "remove first worker". Also, do we really want the latter?

Reply:
> I don't really like the dual meaning of "force" as both "delete last shard placement" and "remove first worker".

I think this makes sense, using the same flag would be confusing.

> Also, do we really want the former? Could users achieve that through manual deletion?

Yes, we can achieve that via manual shard deletion like: 

```SQL
-- remove all placements on the given node
 with shards_on_node AS (DELETE FROM pg_dist_shard_placement WHERE nodename = 'localhost' AND  nodeport = 9700 RETURNING shardid) DELETE FROM pg_dist_shard WHERE shardid IN (SELECT shardid FROM shards_on_node);
-- then disable
select citus_disable_node('localhost', 9700);
```

So, I'll update the tests, but it might take some time :)